src/passes/RemoveUnusedBrs.cpp

```diff
@@ -906,6 +906,18 @@ struct RemoveUnusedBrs : public WalkerPass<PostWalker<RemoveUnusedBrs>> {
         // further optimizations after this, and those optimizations might even
         // benefit from this improvement.
         auto glb = Type::getGreatestLowerBound(curr->castType, refType);
+        if (curr->op == BrOnCastFail) {
+          // BrOnCastFail sends the input type, with adjusted nullability. The
+          // input heap type makes sense for the branch target, and we will not
+          // change it anyhow, but we need to be careful with nullability: if
+          // the cast type was nullable, then we were sending a non-nullable
+          // value to the branch, and if we refined the cast type to non-
+          // nullable, we would no longer be doing that. In other words, we must
+          // not refine the nullability, as that would *un*refine the send type.
+          if (curr->castType.isNullable() && glb.isNonNullable()) {
+            glb = glb.with(Nullable);
+          }
```

Comment:
We probably need to make the same fix in `BrOn::finalize`, which also sets `castType` to be a GLB.

Reply:
Hmm, I think it works out there, since we don't store the input type. So the input type is equal to the child reference's type, which we compute the GLB on.

Reply:
Refining the child to be non-nullable and then refinalizing will cause the cast type to become non-nullable and the sent type to become nullable. Maybe the later refinalization of the branch target an all its ancestors will handle that correctly, though?

Reply:
I don't think it can get that far:
```wat
(module
 (func $br-on-cast-fail-glb-sent (param $ref (ref any)) (result (ref any))
  (block $block (result (ref any))
   (drop
    (block (result i31ref)
     (br_on_cast_fail $block anyref i31ref
      (local.get $ref)
     )
    )
   )
   (unreachable)
  )
 )
)
```
In the input wat we have input type anyref and cast type i31ref, so the sent type is non-nullable. But in the IR that turns into
```
     (br_on_cast_fail $block (ref any) (ref i31)
```
The input type is now non-nullable because we don't store the declared input type. Then it is ok that we refine the cast type: we are still sending non-nullable.

Reply:
To clarify, the reason this pass *can* hit the problem is that it doesn't look at `curr->ref->type`, but at the fallthrough value. So it is looking at something more than what is in the wasm validation rules.

Reply:
Ok, right, because if we refine the input to `br_on_cast_fail` to be non-nullable, then the sent type would necessarily stay non-nullable. Thanks!